add desiutil.healpix

sbailey opened the pull request:
This PR adds `desiutil.healpix` for working with healpix, including
  * mapping ra,dec to healpixels (hpix) and unique pixel identifiers (upix)
  * converting to/from nside,healpix and unique pixels
  * partitioning ra,dec arrays into unique pixels of varying nsides with a maximum number of elements each
  * finding which upix are covered by ra,dec given an array of potential available upix

These utilities are in the context of desiutil/desispec#2510, to provide a way to use adaptively sized nested healpix to handle varying densities on the sky.  They are provided here in desiutil to make them available to more packages without introducing dependency loops.

This includes tests with 100% coverage.

Note: `desiutil.healpix.radec2hpix` is redundant with `desimodel.footprint.radec2pix`, but this module provides additional functionality and a better dependency location.

moustakas commented:
IMHO, `healpy` is so integral to the DESI ecosystem that promoting it to a required dependency in `desiutil` should not be controversial.

sbailey commented:
From Ben:
> Before anything else, is this PR on a critical path and needs to be merged with a subsequent tag, say in the next couple of days?

It is not on the critical path.  It would be nice to get it merged so that it can be used from main for other development work, but it doesn't need an urgent tag.

From John:
> IMHO, healpy is so integral to the DESI ecosystem that promoting it to a required dependency in desiutil should not be controversial.

The concept of healpixels is integral to the desi data products, but for many cases end-users do not need to use healpy code itself.  e.g. HEALPIX is kept as a column in the redshift catalogs such that users can lookup the spectra file locations without having to do a (ra,dec) -> healpix calculation using healpy.

I suggest that we finalize and merge this PR without changing setup.cfg (i.e. keeping healpy as an optional dependency) and then deal with any dependency specific changes in a separate PR dedicated to that if we decide we need it.  At NERSC we already have healpy installed, and the [tutorials setup guide](https://github.com/desihub/tutorials/blob/main/01_getting_started/00_Setup.md) also includes explicitly installing healpy before desiutil, so in the short term I don't think there is any pratical downside to leaving it as optional in setup.cfg.

sbailey commented on py/desiutil/healpix.py:
```diff
@@ -0,0 +1,194 @@
+================
+
+Utilities for working with healpix tiling, including unique pixel
+encoding both nside,healpix into a single integer:
```
done

sbailey commented on py/desiutil/healpix.py:
```diff
@@ -0,0 +1,194 @@
+    Args:
+        upix (int or array of int): encoded unique pixel(s)
+
+    Returns (nsides, healpixels)
```
The previous output was readable but didn't follow the standard formatting; I have updated this as suggested.

sbailey commented on py/desiutil/healpix.py:
```diff
@@ -0,0 +1,194 @@
+        if np.any(too_many):
+            ii = np.isin(upix, upix_values[too_many])
+            nside = 2**order
+            print(f'{np.sum(too_many)}/{len(too_many)} upix with {np.sum(ii)} targets are too dense; splitting those from nside={nside//2} to {nside}')
```
In this case I purposefully used a print instead of log.info so that the module would be completely standalone and could be copied into user-code without needing any other user code.  This is in the spirit of making DESI file outputs usable completely independently of DESI code, and this `find_upix` code is slightly tricky that a user might want to copy without bringing in the full desiutil.

I could have made this a generic python logger, but that wouldn't have standard desi formatting etc.  Instead, I updated this to have a verbose option so that you only get messages if you want, and added a section to the top docstring explaining that this is standalone and can be copied as long as you follow LICENSE.rst.

weaverba137 replied:
This is fine, but I caught a typo, see below.

sbailey commented on py/desiutil/healpix.py:
```diff
@@ -0,0 +1,194 @@
+    scalar_input = np.isscalar(ra)
+    ra = np.atleast_1d(ra)
+    dec = np.atleast_1d(dec)
+    assert len(ra) == len(dec)
```
done (and tests added to check that)

sbailey commented on py/desiutil/healpix.py:
```diff
@@ -0,0 +1,194 @@
+    # so that radec2hpix is only calculated once (slow).
+
+    available_nside = np.unique(upix2hpix(available_upix)[0])
+    assert available_nside[0] >= 1
```
This assert should never fail, but I added it as a safety check while debugging because otherwise the code could end up in an infinite loop due to `nside = nside // 2` always being stuck at 0.

sbailey commented:
I believe I have addressed all of the review comments, though in the case of logging I explained the reasoning for using print instead.  Please re-check.

sbailey commented:
fixed typo and added license and encoding comment lines.

weaverba137 commented:
OK to merge when ready.
